src/qubes_fwupdmgr.py

```diff
@@ -0,0 +1,502 @@
+    ],
+    "Help": [
+        {
+            "-h --help": "Show help options"
+        }
+    ]
+}
+
+
+class QubesFwupdmgr:
+    def _download_metadata(self):
+        """Initialize downloading metadata files."""
+        cmd_metadata = [
+            FWUPD_DOM0_UPDATE,
+            "--metadata"
+        ]
+        p = subprocess.Popen(cmd_metadata)
+        p.wait()
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Metadata update failed")
+        if not os.path.exists(FWUPD_DOM0_METADATA_FILE):
+            raise ValueError("Metadata signature does not exist")
+
+    def refresh_metadata(self):
+        """Updates metadata with downloaded files."""
+        self._download_metadata()
+        cmd_refresh = [
+            "/bin/fwupdmgr",
+            "refresh",
+            FWUPD_DOM0_METADATA_FILE,
+            FWUPD_DOM0_METADATA_SIGNATURE,
+            "lvfs"
+        ]
+        p = subprocess.Popen(
+            cmd_refresh,
+            stdout=subprocess.PIPE
+        )
+        self.output = p.communicate()[0].decode()
+        print(self.output)
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Refresh failed")
+        if not METADATA_REFRESH_REGEX.match(self.output):
+            raise ValueError("Metadata signature does not exist")
+
+    def _get_updates(self):
+        """Gathers infromations about available updates."""
+        cmd_get_updates = [
+            "/bin/fwupdagent",
+            "get-updates"
+        ]
+        p = subprocess.Popen(
+            cmd_get_updates,
+            stdout=subprocess.PIPE
+        )
+        self.updates_info = p.communicate()[0].decode()
+        if p.returncode != 0:
+            raise Exception("fwudp-qubes: Getting available updates failed")
+
+    def _parse_updates_info(self, updates_info):
+        """Creates dictionary and list with informations about updates.
+
+        Keywords argument:
+        updates_info - gathered update informations
+        """
+        self.updates_info_dict = json.loads(updates_info)
+        self.updates_list = [
+            {
+                    "Name": device["Name"],
+                    "Version": device["Version"],
+                    "Releases": [
+                        {
+                            "Version": update["Version"],
+                            "Url": update["Uri"],
+                            "Checksum": update["Checksum"][0],
+                            "Description": update["Description"]
+                        } for update in device["Releases"]
+                    ]
+            } for device in self.updates_info_dict["Devices"]
+        ]
+
+    def _download_firmware_updates(self, url, sha):
+        """Initializes downloading firmware upadate archive.
+
+        Keywords arguments:
+        url -- url path to firmware upadate archive
+        sha -- SHA1 checksum of the firmware update archive
+        """
+        name = url.replace("https://fwupd.org/downloads/", "")
```

Comment:
Maybe this should be a constant as well? I can see it being used a few times.

Reply:
You're right. Done.